runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/Engine.java

```diff
@@ -481,6 +488,45 @@ public long[][] histogramIds()
         return worker.histogramIds();
     }
 
+    public int readEvent(
+        MessageConsumer handler,
+        int messageCountLimit)
+    {
+        int messagesRead = 0;
+        boolean empty = false;
+        while (!empty && messagesRead < messageCountLimit)
+        {
+            int eventCount = 0;
+            minWorkerIndex = 0;
+            minTimeStamp = Long.MAX_VALUE;
+            for (int j = 0; j < workers.size(); j++)
+            {
+                final int workerIndex = j;
+                int eventPeeked = workers.get(workerIndex).peekEvent((m, b, i, l) ->
+                {
+                    eventRO.wrap(b, i, i + l);
+                    if (eventRO.timestamp() < minTimeStamp)
+                    {
+                        minTimeStamp = eventRO.timestamp();
+                        minWorkerIndex = workerIndex;
+                    }
+                });
+                eventCount += eventPeeked;
+            }
+            empty = eventCount == 0;
+            if (!empty)
+            {
+                messagesRead += workers.get(minWorkerIndex).readEvent(handler, 1);
+            }
+        }
+        return messagesRead;
+    }
```

Comment:
I believe if we want to make sure the events are returned in the proper order, we must re-peek the workers in every iteration e.g. if worker A has 1 2, worker B has 3, worker C has 4, then in the first iteration we return 1 from worker A and then in the next iteration we need to return 2 from worker A again, but we miss that if we don't re-peek.

I think in terms of performance it's the fastest if we just pick an event round-robin style without giving any guarantees around ordering. This would be a slight issue for the stdout exporter occasionally displaying events in the wrong order (but with the correct timestamp), and probably a non-issue for other exporters, where the timestamp matters but most probably the events are not expected to be sent in the exact order in the first place.